src/v/transform/transform_manager.cc

```diff
@@ -0,0 +1,386 @@
+          std::hash<model::transform_id>,
+          std::equal_to<>>>>;
+
+public:
+    const entry_t& insert(
+      std::unique_ptr<processor> processor,
+      ss::lw_shared_ptr<transform::probe> probe) {
+        auto [it, inserted] = _underlying.emplace(
+          std::move(processor), std::move(probe));
+        vassert(inserted, "invalid transform processor management");
+        return *it;
+    }
+
+    ss::lw_shared_ptr<probe> get_or_create_probe(
+      model::transform_id id, const model::transform_metadata& meta) {
+        auto& id_index = _underlying.get<id_index_t>();
+        auto id_it = id_index.find(id);
+        if (id_it == id_index.end()) {
+            auto probe = ss::make_lw_shared<transform::probe>();
+            probe->setup_metrics(meta.name);
+            return probe;
+        }
+        return id_it->probe();
+    }
+
+    bool contains(model::transform_id id, const model::ntp& ntp) {
+        auto& by_key = _underlying.get<key_index_t>();
+        return by_key.contains(std::make_tuple(id, ntp));
+    }
+
+    const entry_t* get_or_null(model::transform_id id, const model::ntp& ntp) {
+        auto& by_key = _underlying.get<key_index_t>();
+        auto it = by_key.find(std::make_tuple(id, ntp));
+        if (it == by_key.end()) {
+            return nullptr;
+        }
+        return std::addressof(*it);
+    }
+
+    ss::future<> clear() {
+        co_await ss::parallel_for_each(
+          _underlying.begin(),
+          _underlying.end(),
+          // NOLINTNEXTLINE(cppcoreguidelines-avoid-reference-coroutine-parameters)
+          [](auto& e) { return e.processor()->stop(); });
+        _underlying.clear();
+    }
+
+    ss::future<> erase_by_id(model::transform_id id) {
+        auto& by_id = _underlying.get<id_index_t>();
+        auto range = by_id.equal_range(id);
+        co_await ss::parallel_for_each(
+          range.first,
+          range.second,
+          // NOLINTNEXTLINE(cppcoreguidelines-avoid-reference-coroutine-parameters)
+          [](auto& e) { return e.processor()->stop(); });
+        by_id.erase(range.first, range.second);
+    }
+
+    // Clear our all the transforms with a given ntp and return all the IDs that
+    // no longer exist.
+    ss::future<> erase_by_ntp(model::ntp ntp) {
+        auto& by_ntp = _underlying.get<ntp_index_t>();
+        auto range = by_ntp.equal_range(ntp);
+        co_await ss::parallel_for_each(
+          range.first,
+          range.second,
+          // NOLINTNEXTLINE(cppcoreguidelines-avoid-reference-coroutine-parameters)
+          [](auto& e) { return e.processor()->stop(); });
+        by_ntp.erase(range.first, range.second);
+    }
+
+private:
+    underlying_t _underlying;
+};
+
+manager::manager(
+  std::unique_ptr<registry> r, std::unique_ptr<processor_factory> f)
+  : _queue([](const std::exception_ptr& ex) {
+      vlog(tlog.error, "unexpected transform manager error: {}", ex);
```

Comment:
should this ignore common exceptions like abort requested, gate closed, etc... ?

Reply:
Good question. Right now none of those exceptions should ever be surfaced, so I think no.

I feel like blanket ignoring some of those exceptions can be a little scary. Part of the hope of managing the async lifecycle in the queue class is that so there is only a single place that needs to handle this.